snapcraft/internal/pluginhandler/stage_package_grammar/_processor.py

```diff
@@ -0,0 +1,138 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+
+from ._errors import StagePackageSyntaxError
+
+
+def process_grammar(grammar, project_options, ubuntu):
+    """Process stage packages grammar and extract packages to actually stage.
+
+    Arguments:
+        grammar: List containing stage-packages grammar.
+        project_options: Instance of ProjectOptions to use to determine stage
+                         packages.
+        ubuntu: repo.Ubuntu instance used for checking package validity.
+
+    Returns:
+        Set of packages to stage.
+    """
+
+    from ._on import OnStatement
+    from ._try import TryStatement
+
+    packages = set()
+    on_clause_pattern = re.compile(r'\Aon\s+')
+    try_clause_pattern = re.compile(r'\Atry\Z')
+    else_clause_pattern = re.compile(r'\Aelse\Z')
+    else_fail_pattern = re.compile(r'\Aelse\s+fail\Z')
+    statements = _StatementCollection()
+    statement = None
+
+    for section in grammar:
+        try:
+            if else_fail_pattern.match(section):
+                _handle_else(statement, None)
+            else:
+                # If it's just a string, add it to the package set.
+                packages.add(section)
+        except TypeError:
+            # Alright, it wasn't an unconditional stage package, so it must be
+            # a dict.
+            for key, value in section.items():
+                if on_clause_pattern.match(key):
+                    # We've come across the begining of an 'on' statement.
+                    # That means any previous statement we found is complete.
+                    statements.add(statement)
```

Comment:
Let's use this as an example:

```
- on amd64:
  - foo
- try:
  - bar
- else:
  - baz
```
`process_grammar()` is called with this entire thing, and it walks through the top level list (`on`, `try`, `else`).

1. **Parsing `on amd64`**
    Since `statement` is `None` initially, yes, when it hits the `on` it adds `None` to the `statements` collection (which will [ignore it](https://github.com/snapcore/snapcraft/pull/1059/files#diff-68c6d965d4ff21ecd61544afbc4bb051R119)). Then it creates a new `OnStatement` and holds onto it.
2. **Parsing `try`**
    Since we [hit a new statement](https://github.com/snapcore/snapcraft/pull/1059/files#diff-68c6d965d4ff21ecd61544afbc4bb051R68) we know the `OnStatement` we were holding onto is complete, so it's added to the collection. It then creates a new `TryStatement` and holds onto it.
3. **Parsing `else`**
    `else` clauses mean nothing by themselves; they [must be attached to an `OnStatement` or `TryStatement`](https://github.com/snapcore/snapcraft/pull/1059/files#diff-68c6d965d4ff21ecd61544afbc4bb051R95), otherwise their presence is an error. Since we already have a `TryStatement`, it gets attached to that.
4. **Parsing is complete**
    But we're still holding onto that `TryStatement`. [Add it to the collection](https://github.com/snapcore/snapcraft/pull/1059/files#diff-68c6d965d4ff21ecd61544afbc4bb051R77).
5. **[Process all statements in collection](https://github.com/snapcore/snapcraft/pull/1059/files#diff-68c6d965d4ff21ecd61544afbc4bb051R78)**